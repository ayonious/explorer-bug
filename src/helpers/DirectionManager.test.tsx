/**
 * @jest-environment jsdom
 */

import {
  moveForward,
  turnLeft,
  turnRight,
  getNewGrid,
  getNewScore,
  robotPosToFlightStyle,
  ROW_LIMIT,
  COL_LIMIT,
} from './DirectionManager';

describe('Robot Movement Tests', () => {
  it('should handle all basic movement combinations', () => {
    [0, 1, 2, 3].forEach((direction) => {
      [moveForward, turnLeft, turnRight].forEach((turnFn) => {
        expect(
          turnFn({
            col: 1,
            row: 2,
            direction,
          })
        ).toMatchSnapshot();
      });
    });
  });

  it('should prevent robot from moving outside grid boundaries', () => {
    // Test upper boundary
    const upState = {
      col: 1,
      row: 0,
      direction: 0, // Facing up
    };
    const upResult = moveForward(upState);
    expect(upResult.row).toBe(0); // Should not move up

    // Test lower boundary
    const downState = {
      col: 1,
      row: ROW_LIMIT - 1,
      direction: 2, // Facing down
    };
    const downResult = moveForward(downState);
    expect(downResult.row).toBe(ROW_LIMIT - 1); // Should not move down

    // Test left boundary
    const leftState = {
      col: 0,
      row: 1,
      direction: 3, // Facing left
    };
    const leftResult = moveForward(leftState);
    expect(leftResult.col).toBe(0); // Should not move left

    // Test right boundary
    const rightState = {
      col: COL_LIMIT - 1,
      row: 1,
      direction: 1, // Facing right
    };
    const rightResult = moveForward(rightState);
    expect(rightResult.col).toBe(COL_LIMIT - 1); // Should not move right
  });

  it('should correctly rotate through all directions', () => {
    let state = { col: 1, row: 1, direction: 0 };
    
    // Test full rotation to the right
    state = turnRight({ ...state });
    expect(state.direction).toBe(1);
    state = turnRight({ ...state });
    expect(state.direction).toBe(2);
    state = turnRight({ ...state });
    expect(state.direction).toBe(3);
    state = turnRight({ ...state });
    expect(state.direction).toBe(0);

    // Reset state
    state = { col: 1, row: 1, direction: 0 };

    // Test full rotation to the left
    state = turnLeft({ ...state });
    expect(state.direction).toBe(3);
    state = turnLeft({ ...state });
    expect(state.direction).toBe(2);
    state = turnLeft({ ...state });
    expect(state.direction).toBe(1);
    state = turnLeft({ ...state });
    expect(state.direction).toBe(0);
  });
});

describe('Grid and Score Management Tests', () => {
  it('should correctly update grid after movement', () => {
    const initialGrid = [
      [1, 0, 1],
      [0, 1, 0],
      [1, 1, 1],
    ];

    const robotState = {
      col: 1,
      row: 1,
      direction: 0,
    };

    const newGrid = getNewGrid({ ...robotState }, JSON.parse(JSON.stringify(initialGrid)));
    expect(newGrid[1][1]).toBe(0); // Food should be eaten
  });

  it('should correctly calculate score changes', () => {
    const grid = [
      [1, 0, 1],
      [0, 1, 0],
      [1, 1, 1],
    ];

    // Test score increase when food is eaten
    expect(
      getNewScore(
        { col: 1, row: 1, direction: 0 },
        10,
        grid
      )
    ).toBe(11);

    // Test score remains same when no food
    expect(
      getNewScore(
        { col: 0, row: 1, direction: 0 },
        10,
        grid
      )
    ).toBe(10);
  });

  it('should handle empty and full grids', () => {
    const emptyGrid = [
      [0, 0, 0],
      [0, 0, 0],
      [0, 0, 0],
    ];

    const fullGrid = [
      [1, 1, 1],
      [1, 1, 1],
      [1, 1, 1],
    ];

    const robotState = {
      col: 1,
      row: 1,
      direction: 0,
    };

    // Empty grid should remain unchanged after deep clone
    const newEmptyGrid = getNewGrid({ ...robotState }, JSON.parse(JSON.stringify(emptyGrid)));
    expect(newEmptyGrid).toEqual(emptyGrid);
    
    // Full grid should change at robot position
    const newFullGrid = getNewGrid({ ...robotState }, JSON.parse(JSON.stringify(fullGrid)));
    expect(newFullGrid[1][1]).toBe(0);
  });
});

describe('Robot Style Tests', () => {
  it('should return correct rotation styles for all directions', () => {
    const robotState = { col: 2, row: 3, direction: 0 };
    
    // Test all directions
    [0, 1, 2, 3].forEach((direction) => {
      expect(
        robotPosToFlightStyle({
          ...robotState,
          direction,
        })
      ).toMatchSnapshot();
    });
  });

  it('should handle edge positions', () => {
    // Test corners
    const corners = [
      { col: 0, row: 0 },
      { col: COL_LIMIT - 1, row: 0 },
      { col: 0, row: ROW_LIMIT - 1 },
      { col: COL_LIMIT - 1, row: ROW_LIMIT - 1 },
    ];

    corners.forEach(({ col, row }) => {
      expect(
        robotPosToFlightStyle({
          col,
          row,
          direction: 0,
        })
      ).toMatchSnapshot();
    });
  });
});
