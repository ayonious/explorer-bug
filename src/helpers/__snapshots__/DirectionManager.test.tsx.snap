// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movements 1`] = `
{
  "col": 2,
  "direction": 0,
  "row": 2,
}
`;

exports[`Movements 2`] = `
{
  "col": 1,
  "direction": 3,
  "row": 2,
}
`;

exports[`Movements 3`] = `
{
  "col": 1,
  "direction": 1,
  "row": 2,
}
`;

exports[`Movements 4`] = `
{
  "col": 1,
  "direction": 1,
  "row": 3,
}
`;

exports[`Movements 5`] = `
{
  "col": 1,
  "direction": 0,
  "row": 2,
}
`;

exports[`Movements 6`] = `
{
  "col": 1,
  "direction": 2,
  "row": 2,
}
`;

exports[`Movements 7`] = `
{
  "col": 0,
  "direction": 2,
  "row": 2,
}
`;

exports[`Movements 8`] = `
{
  "col": 1,
  "direction": 1,
  "row": 2,
}
`;

exports[`Movements 9`] = `
{
  "col": 1,
  "direction": 3,
  "row": 2,
}
`;

exports[`Movements 10`] = `
{
  "col": 1,
  "direction": 3,
  "row": 1,
}
`;

exports[`Movements 11`] = `
{
  "col": 1,
  "direction": 2,
  "row": 2,
}
`;

exports[`Movements 12`] = `
{
  "col": 1,
  "direction": 0,
  "row": 2,
}
`;

exports[`RobotState 1`] = `
{
  "transform": "rotate(90deg)",
}
`;

exports[`RobotState 2`] = `
{
  "transform": "rotate(180deg)",
}
`;

exports[`RobotState 3`] = `
{
  "transform": "rotate(270deg)",
}
`;

exports[`RobotState 4`] = `
{
  "transform": "",
}
`;
